### ISSUE TITLE
errorhandling

### ISSUE THREAD

**Issue by StefanHoutzager** [Lucifier129/farrow#16]
Interesting work. Have you looked into "type safe errorhandling"? For example https://dev.to/_gdelgado/type-safe-error-handling-in-typescript-1p4n explains it a bit.

**Comment by Lucifier129** [comment]
Yeah, I have learned Haskell and Rust, but haven't seen the article above before.

**Comment by uinz** [comment]
To be honest, I tend to confuse errors and exceptions at work，but I think they are very different

**Comment by Lucifier129** [comment]
@uinz Maybe thinking with the inhabitants of type can help.

- `zero` inhabitants type like` `Void` or `never` or `panic`, means halt, no program can run.
- `1 + a` inhabitants type like `Maybe` in `Haskell`, or `Option` in `Rust`,  encode the empty or defaults with a specific type
- `a + b` inhabitants types like `Either` in `Haskell`, or `Result` in `Rust`, encode the union of two types.
